### PR TITLE
Update for PHPCS 3.x and WPCS 0.14.1

### LINKED ISSUE #110: Restrict defining deprecated define variables
Separated from https://github.com/WPTRT/WordPress-Coding-Standards/pull/30#issuecomment-274225355

```
define('HEADER_IMAGE_WIDTH',1200);
define('HEADER_IMAGE_HEIGHT',600);
```
This type of code and similar should be caught by the sniff I believe. This code wont give `Constant already defined` error, since this was the recommended method used to override Custom Header before 3.4.

### LINKED ISSUE #139: Support PHPCS 3.x
Once WPCS 0.13.0 has been released next week, I would like to propose to do a back-merge again and to update the existing code within the TRTCS to be in line with WPCS.

This includes support for PHPCS 3.x, using namespaces and more.

### LINKED ISSUE #144: PHP 7.2 Deprecated
Apparently VVV upgraded their PHP to 7.2 and getting the following deprecated

<b>Deprecated</b>:  Function create_function() is deprecated in <b>/srv/www/getbenonit/public_html/wp-content/plugins/theme-sniffer/vendor/wp-coding-standards/wpcs/WordPress/Sniffs/VIP/ValidatedSanitizedInputSniff.php</b> on line <b>93</b><br />

I was told to open an issue here regarding the deprecated. 

as @dingo-d mention ed follow

This seems like an upstream issue with the WordPress Coding Standards (vendor/wp-coding-standards/wpcs/WordPress/Sniffs/VIP/ValidatedSanitizedInputSniff.php).

### PR DESCRIPTION
**Pull request by jrfnl**
## TL;DR
This PR does a back-merge of WPCS to WPTRT based on WPCS `master`/`0.14.1`.

It then applies a number of commits to make the Theme sniffs consistent with developments in WPCS upstream.

Lastly, it updates the sniffs to be compatible with PHPCS 3.x.

### Merge branch 'master' into WPTRT/feature/update-forPHPCS-3.x-WPCS-0.14

Note: For now, I've used the `master` branch of WPCS as there are some issues which still need to be addressed in WPCS `develop`. Once those have been sorted, `develop` should be merged in as it contains significant bug fixes,

Fixes #144

### DeprecatedWPConstants: Remove sniff in favour of WPCS sniff

WPCS 0.14.0 introduced a `WP.DiscouragedConstants` sniff which covers all the constants which were covered by the `DeprecatedWPConstants` sniff and does so better than the original sniff.

Removed the `DeprecatedWPConstants` sniff in favour of the new upstream sniff and changing the error level from `warning` to `error` from the ruleset maintains the existing behaviour.

Ref: WordPress-Coding-Standards/WordPress-Coding-Standards#1089

Closes #140 
Fixes #110

### Leverage the WordPress_Sniff class for all sniffs

Most WPCS sniffs extend the WordPress_Sniff class or one of the abstract WP sniffs. However there were some which so far didn't.

This PR changes that whenever possible.

The affected sniffs do not necessarily need to use the WordPress_Sniff class as they don't use any of the utility methods or properties contained therein, however it will make future development slightly more intuitive as (nearly) all sniffs now follow the same pattern.

This same change has been made upstream in WordPress-Coding-Standards/WordPress-Coding-Standards#967

### Documentation: Fix two outdated & incorrect code samples

`eval` can not be sniffed for as a function as it is a language construct, not a function, so this was misleading example code.

This same change has been made upstream in WordPress-Coding-Standards/WordPress-Coding-Standards#812

### PHPCS 3.x compat: Ruleset: Make sure the complete sniff is included, not just an errorcode

In PHPCS 2.x when a sniff was referenced by errorcode, this automatically included the whole sniff.
As of PHPCS 3.x it is possible to selectively include just and only a specific errorcode.

This allows for cross-version compat and ensures that if/when new errorcodes are added to the sniffs, they will be included as well.

### PHPCS 3.x compat: Fix all references to upstream PHPCS classes

* Add use statements for all PHPCS class references.
* Where possible use the PHPCS 3.x class name as a use alias to make the future change of dropping 2.x support easier.

### PHPCS 3.x compat: Add autoload directive for the alias file

The `<autoload>` ruleset directive was added in PHPCS 3.0.1 in response to issue squizlabs/PHP_CodeSniffer#1469

### Clean up: Remove `// end` comments

The upstream CommentedOutCode sniff has been updated and become more accurate and now throws a number of warnings for // End ... comments.

As the rule regarding end comments has been removed from the handbook anyway, we may as well remove them all and get rid of the warnings.

This same change has been made upstream in WordPress-Coding-Standards/WordPress-Coding-Standards#1362

---- 
Fixes #139 

/cc @dingo-d

### CODE REVIEW

**Comment by jrfnl** [comment]
I've added an additional commit to be in line with the latest PR merged into WPCS:

### Clean up: Remove `// end` comments

The upstream CommentedOutCode sniff has been updated and become more accurate and now throws a number of warnings for // End ... comments.

As the rule regarding end comments has been removed from the handbook anyway, we may as well remove them all and get rid of the warnings.

This same change has been made upstream in WordPress-Coding-Standards/WordPress-Coding-Standards#1362

File: WordPress/Tests/CodeAnalysis/EmptyStatementUnitTest.php [hunk 1]
```diff
@@ -7,34 +7,27 @@
  * @license https://opensource.org/licenses/MIT MIT
  */
 
+namespace WordPress\Tests\CodeAnalysis;
```
Comment (grappler) [inline WordPress/Tests/CodeAnalysis/EmptyStatementUnitTest.php]: This file seems to be incorrectly renamed.

Comment (jrfnl) [reply WordPress/Tests/CodeAnalysis/EmptyStatementUnitTest.php]: That's just Git being weird, nothing to worry about.

**Comment by jrfnl** [comment]
> I wonder if we can get the git rename fixed.

That's just an artifact of the fact that we do a back-merge here. It may look a little weird when viewing the Git history, but I can't partially back-merge. In WPCS the history is unbroken.

As there is a [proposal to make this repo stand-alone](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/578#issuecomment-392082401), this may soon not be relevant anymore anyway.

**Comment by grappler** [comment]
Thank you for working on this!
